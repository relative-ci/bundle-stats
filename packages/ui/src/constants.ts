import { FILE_TYPES, SECTIONS } from '@bundle-stats/utils';

export const URLS = {
  OVERVIEW: '/',
  ASSETS: '/assets',
  MODULES: '/modules',
  PACKAGES: '/packages',
};

export const SECTION_URLS = {
  [SECTIONS.TOTALS]: URLS.OVERVIEW,
  [SECTIONS.ASSETS]: URLS.ASSETS,
  [SECTIONS.MODULES]: URLS.MODULES,
  [SECTIONS.PACKAGES]: URLS.PACKAGES,
};

export const ASSETS_SIZES_FILE_TYPE_MAP = FILE_TYPES.reduce(
  (agg, fileType) => ({
    ...agg,
    [`webpack.sizes.totalSizeByType${fileType}`]: fileType,
  }),
  {},
);

export const METRICS_WEBPACK_GENERAL = [
  'webpack.totalSizeByTypeALL',
  'webpack.totalInitialSizeJS',
  'webpack.totalInitialSizeCSS',
  'webpack.cacheInvalidation',
];
export const METRICS_WEBPACK_ASSETS = ['webpack.assetCount', 'webpack.chunkCount'];
export const METRICS_WEBPACK_MODULES = [
  'webpack.moduleCount',
  'webpack.duplicateModulesCount',
  'webpack.duplicateCode',
];
export const METRICS_WEBPACK_PACKAGES = ['webpack.packageCount', 'webpack.duplicatePackagesCount'];

export const SORT = {
  ASC: 'asc',
  DESC: 'desc',
};
