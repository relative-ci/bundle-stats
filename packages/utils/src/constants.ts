export enum Source {
  webpack = 'webpack',
  lighthouse = 'lighthouse',
  browsertime = 'browsertime',
}

export type SourceData = Record<string, unknown>;

export enum InsightType {
  ERROR = 'error',
  INFO = 'info',
  WARNING = 'warning',
}

/**
 * Generic metric type names
 */
export enum MetricTypes {
  Numeric = 'METRIC_TYPE_NUMERIC',
  Score = 'METRIC_TYPE_SCORE',
  FileSize = 'METRIC_TYPE_FILE_SIZE',
  Duration = 'METRIC_TYPE_DURATION',
  Percentage = 'METRIC_TYPE_PERCENTAGE',
}

/**
 * Metric type base configuration
 */
export interface MetricTypeConfig {
  /**
   * Increased values are regressions, decreased values are improvements
   */
  biggerIsBetter?: boolean | null;
  /**
   * Value formatter
   */
  formatter: (value: number | null) => string;
}

/**
 * Source metric configuration
 */
export interface MetricConfig {
  label: string;
  type: MetricTypes;
  description?: string;
  /**
   * Documentation URL
   */
  url?: string;
  /**
   * Positive delta values are regressions and negative delta values are improvements
   * overrides MetricTypeConfig['biggerIsBetter']
   */
  biggerIsBetter?: MetricTypeConfig['biggerIsBetter'];
  /**
   * Do not show the delta
   */
  skipDelta?: boolean;
}

export type Metric = MetricTypeConfig & MetricConfig;

export enum DeltaType {
  HIGH_NEGATIVE = 'HIGH_NEGATIVE',
  NEGATIVE = 'NEGATIVE',
  LOW_NEGATIVE = 'LOW_NEGATIVE',
  NO_CHANGE = 'NO_CHANGE',
  LOW_POSITIVE = 'LOW_POSITIVE',
  POSITIVE = 'POSITIVE',
  HIGH_POSITIVE = 'HIGH_POSITIVE',
}

export type MetricRunInfoDeltaType = `${DeltaType}`;

export interface MetricRun {
  value: number;
}

export interface MetricRunDelta {
  delta: number;
  deltaPercentage: number;
}

export interface MetricRunInfo {
  value: number;
  displayValue: string;
  delta?: number;
  deltaPercentage?: number;
  deltaType?: MetricRunInfoDeltaType;
  displayDelta?: string;
  displayDeltaPercentage?: string;
  regression?: boolean | null;
}

export type JobSection<T = object> = Record<Source, T>;

export enum JobSectionId {
  meta = 'meta',
  insights = 'insights',
  summary = 'summary',
  metrics = 'metrics',
  rawData = 'rawData',
}

export interface JobSummaryItem {
  baseline: number;
  current: number;
}

export interface JobInsight<T = object> {
  type: InsightType;
  data: T & { text?: string };
}

export interface JobInsightAssetsSizeTotalData {
  md: string;
  text: string;
  info: MetricRunInfo;
}
export type JobInsightDuplicatePackagesData = Record<string, Array<string>>;

export interface JobInsightDuplicatePackagesV3Data {
  packages: Record<string, Array<string>>;
}

export interface JobInsightNewPackagesData {
  packages: Array<string>;
}

export type JobSummarySource = Record<string, JobSummaryItem>;
export type JobSummary = JobSection<JobSummarySource>;

export interface JobInsights {
  [Source.webpack]: {
    assetsSizeTotal: JobInsight<JobInsightAssetsSizeTotalData>;
    duplicatePackages?: JobInsight<JobInsightDuplicatePackagesData>;
    duplicatePackagesV3?: JobInsight<JobInsightDuplicatePackagesV3Data>;
    newPackages?: JobInsight<JobInsightNewPackagesData>;
  };
}

export interface JobInsightsInfo {
  duplicatePackages?: JobInsights['webpack']['duplicatePackagesV3'];
  newPackages?: JobInsights['webpack']['newPackages'];
}

export type JobMetricsSource = Record<string, MetricRun | Record<string, MetricRun>>;
export type JobMetrics = JobSection<JobMetricsSource>;

export interface JobData {
  [JobSectionId.meta]?: JobSection;
  [JobSectionId.insights]?: JobInsights;
  [JobSectionId.summary]?: JobSummary;
  [JobSectionId.metrics]?: JobMetrics;
  [JobSectionId.rawData]?: JobSection;
}

export interface Job extends JobData {
  internalBuildNumber: number;
  // @TODO(v5): Remove or move to report
  label: string;
}

export interface LighthouseSource {
  lighthouseVersion: string;
  fetchTime: string;
  requestedUrl: string;
  categories: { [key: string]: { score: number } };
  audits: { [key: string]: { score: number; numericValue: number } };
}

export interface LighthouseMetricsCategoryScores {
  performanceScore: MetricRun;
  accessibilityScore: MetricRun;
  bestPracticesScore: MetricRun;
  seoScore: MetricRun;
  pwaScore: MetricRun;
}

export interface LighthouseMetricsScore {
  score: MetricRun;
}

export interface LighthouseMetricsAudits {
  speedIndex: MetricRun;
  firstMeaningfulPaint: MetricRun;
  timeToFirstByte: MetricRun;
  firstInteractive: MetricRun;
  totalByteWeight: MetricRun;
  domSize: MetricRun;
}

export interface BrowsertimeSourceMetric {
  median: number;
}

export interface BrowsertimeSource {
  info?: {
    browsertime: {
      version: string;
    };
    timestamp: string;
    url: string;
  };
  statistics?: {
    timings: {
      firstPaint: BrowsertimeSourceMetric;
      fullyLoaded: BrowsertimeSourceMetric;
      pageTimings: {
        backEndTime: BrowsertimeSourceMetric;
        domContentLoadedTime: BrowsertimeSourceMetric;
        domInteractiveTime: BrowsertimeSourceMetric;
        domainLookupTime: BrowsertimeSourceMetric;
        frontEndTime: BrowsertimeSourceMetric;
        pageDownloadTime: BrowsertimeSourceMetric;
        pageLoadTime: BrowsertimeSourceMetric;
        redirectionTime: BrowsertimeSourceMetric;
        serverConnectionTime: BrowsertimeSourceMetric;
        serverResponseTime: BrowsertimeSourceMetric;
      };
      paintTiming: {
        'first-contentful-paint': BrowsertimeSourceMetric;
      };
      rumSpeedIndex: BrowsertimeSourceMetric;
    };
    visualmetrics: {
      FirstVisualChange: BrowsertimeSourceMetric;
      LastVisualChange: BrowsertimeSourceMetric;
      PerceptualSpeedIndex: BrowsertimeSourceMetric;
      SpeedIndex: BrowsertimeSourceMetric;
      VisualComplete85: BrowsertimeSourceMetric;
      VisualComplete95: BrowsertimeSourceMetric;
      VisualComplete99: BrowsertimeSourceMetric;
    };
  };
}

export interface BrowsertimeMetrics {
  firstPaint: MetricRun;
  fullyLoaded: MetricRun;
  backEndTime: MetricRun;
  domContentLoadedTime: MetricRun;
  domInteractiveTime: MetricRun;
  domainLookupTime: MetricRun;
  frontEndTime: MetricRun;
  pageDownloadTime: MetricRun;
  pageLoadTime: MetricRun;
  redirectionTime: MetricRun;
  serverConnectionTime: MetricRun;
  serverResponseTime: MetricRun;
  firstContentfulPaint: MetricRun;
  rumSpeedIndex: MetricRun;
  firstVisualChange: MetricRun;
  lastVisualChange: MetricRun;
  perceptualSpeedIndex: MetricRun;
  speedIndex: MetricRun;
  visualComplete85: MetricRun;
  visualComplete95: MetricRun;
  visualComplete99: MetricRun;
}
