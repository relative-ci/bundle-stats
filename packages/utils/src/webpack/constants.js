import { Metric } from './types';

export const SECTION_WEBPACK_STATS = 'stats';
export const SECTION_WEBPACK_SIZES = 'sizes';
export const SECTION_WEBPACK_ASSETS = 'assets';
export const SECTION_WEBPACK_MODULES = 'modules';
export const SECTION_WEBPACK_PACKAGES = 'packages';

export const SECTIONS = [
  SECTION_WEBPACK_STATS,
  SECTION_WEBPACK_SIZES,
  SECTION_WEBPACK_ASSETS,
  SECTION_WEBPACK_MODULES,
  SECTION_WEBPACK_PACKAGES,
];

export const SUMMARY_METRIC_PATHS = [
  Metric.BUNDLE_SIZE,
  Metric.INITIAL_SIZE_JS,
  Metric.INITIAL_SIZE_CSS,
  Metric.CACHE_INVALIDATION,
  Metric.MODULE_COUNT,
  Metric.DUPLICATE_MODULES_COUNT,
  Metric.DUPLICATE_CODE,
  Metric.CHUNK_COUNT,
  Metric.ASSET_COUNT,
  Metric.PACKAGE_COUNT,
  Metric.DUPLICATE_PACKAGES_COUNT,
];

const PACKAGE_PREFIX = /(?:node_modules|~)(?:\/\.pnpm)?/;
const PACKAGE_SLUG = /[a-zA-Z0-9]+(?:[-|_|.]+[a-zA-Z0-9]+)*/;
const VERSION = /@[\w|\-|_|.]+/;

// Extract package paths from module path
// https://regex101.com/r/22Leep/6
/* eslint-disable prettier/prettier */
export const MODULE_PATH_PACKAGES = new RegExp(
  [
    // match dependency directory (eg: `node_modules/`, `node_modules/.pnpm/`)
    `(?:${PACKAGE_PREFIX.source}/)`,
    // match package name
    '(?:',
    // match `@organization/` or `@organization+`(pnpm)
    `(?:@${PACKAGE_SLUG.source}[/|+])?`,
    // match github.com+organization+
    `(?:${PACKAGE_SLUG.source}\\+)*`,
    // match package name
    `(?:${PACKAGE_SLUG.source})`,
    // match version
    `(?:${VERSION.source})?`,
    ')',

    // Match pnpm peer dependencies (eg: package-a@version_package-b@version)
    '(?:_',
    `(?:@${PACKAGE_SLUG.source}[/|+])?`,
    `(?:${PACKAGE_SLUG.source})`,
    `(?:@${PACKAGE_SLUG.source})?`,
    ')*',
    '/',
  ].join(''),
  'g',
);
/* eslint-enable prettier/prettier */

// Extract package name from package path
// https://regex101.com/r/tTlU0W/6
/* eslint-disable prettier/prettier */
export const PACKAGE_PATH_NAME = new RegExp(
  [
    `(?:${PACKAGE_PREFIX.source}/)`,
    // match dependency
    '(?:',
    '(',
    `(?:@${PACKAGE_SLUG.source}[/|+])?`,
    `(?:(?:${PACKAGE_SLUG.source}\\+)*)`,
    `(?:${PACKAGE_SLUG.source})`,
    ')',
    `(?:${VERSION.source})?`,
    ')',
    // match peer dependency
    '(?:_',
    '(',
    `(?:@${PACKAGE_SLUG.source}[/|+])?`,
    `(?:${PACKAGE_SLUG.source})`,
    ')',
    `(?:@${PACKAGE_SLUG.source})`,
    ')*',
    '/',
  ].join(''),
  'g',
);
/* eslint-enable prettier/prettier */
