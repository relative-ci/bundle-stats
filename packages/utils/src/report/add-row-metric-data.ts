import get from 'lodash/get';

import { getGlobalMetricType, getMetricRunInfo } from '../utils/metrics';
import { getMetricChanged } from './get-metric-changed';
import * as types from './types';
import { MetricTypes } from '../constants';

/**
 * Add row metric diff data
 */
export const getAddRowMetricData = (metricType?: string) => (row: types.ReportRow) => {
  const { key, runs } = row;

  // Resolve row metric
  // - if the key is a predefined metric, use it
  // - if the key is not matching an existing metric, use the default metricType
  const metric = getGlobalMetricType(
    key,
    typeof metricType === 'string' ? (metricType as MetricTypes) : undefined, // explicit, avoid passing of map cb params
  );
  const { biggerIsBetter, label } = metric;

  const processedRuns = runs.map((run, index) => {
    if (!run || typeof run.value === 'undefined' || run.value === null) {
      return null;
    }

    const metricRunInfo = getMetricRunInfo(
      metric,
      run.value,
      // Get baseline value only if not the latest run
      index + 1 < runs.length ? get(runs, [index + 1, 'value'], 0) : undefined,
    );

    return {
      ...run,
      ...metricRunInfo,
    };
  });

  return {
    ...row,

    // Metric props
    biggerIsBetter,
    label,

    // Row props
    changed: getMetricChanged(runs),

    // Runs
    runs: processedRuns,
  };
};
