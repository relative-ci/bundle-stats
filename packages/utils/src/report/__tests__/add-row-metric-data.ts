import { MetricTypes } from '../../constants';
import { getAddRowMetricData } from '../add-row-metric-data';

describe('report / addRowMetricData', () => {
  test('should add data', () => {
    const rows = [
      {
        key: 'webpack.totalSizeByTypeALL',
        runs: [
          {
            value: 100,
          },
          {
            value: 110,
          },
        ],
      },
      {
        key: 'webpack.duplicatePackagesCount',
        runs: [
          {
            value: 10,
          },
          null,
        ],
      },
      {
        key: 'webpack.moduleCount',
        runs: [
          {
            value: 100,
          },
          {
            value: 100,
          },
        ],
      },
      {
        key: 'webpack.cacheInvalidation',
        runs: [
          null,
          {
            value: 10,
          },
        ],
      },
    ];

    const addRowMetricData = getAddRowMetricData();
    const actual = rows.map(addRowMetricData);

    expect(actual).toMatchSnapshot();
  });

  test('should add data with metric type', () => {
    const rows = [
      {
        key: 'metric1',
        runs: [
          {
            value: 100,
          },
          {
            value: 110,
          },
        ],
      },
      {
        key: 'metric2',
        runs: [
          {
            value: 10,
          },
          null,
        ],
      },
      {
        key: 'metric3',
        runs: [
          null,
          {
            value: 10,
          },
        ],
      },
      {
        key: 'metric4',
        runs: [
          {
            value: 100,
          },
          {
            value: 100,
          },
        ],
      },
    ];

    const addRowMetricData = getAddRowMetricData(MetricTypes.FileSize);
    const actual = rows.map(addRowMetricData);

    expect(actual).toMatchSnapshot();
  });
});
