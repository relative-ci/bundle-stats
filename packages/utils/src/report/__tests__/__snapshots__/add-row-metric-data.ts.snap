// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report / addRowMetricData should add data 1`] = `
[
  {
    "biggerIsBetter": false,
    "changed": true,
    "key": "webpack.totalSizeByTypeALL",
    "label": "Bundle Size",
    "runs": [
      {
        "delta": -10,
        "deltaPercentage": -9.09090909,
        "deltaType": "POSITIVE",
        "displayDelta": "-10B",
        "displayDeltaPercentage": "-9.09%",
        "displayValue": "100B",
        "regression": false,
        "value": 100,
      },
      {
        "displayValue": "110B",
        "value": 110,
      },
    ],
  },
  {
    "biggerIsBetter": false,
    "changed": true,
    "key": "webpack.duplicatePackagesCount",
    "label": "Duplicate Packages",
    "runs": [
      {
        "delta": 10,
        "deltaPercentage": 100,
        "deltaType": "HIGH_NEGATIVE",
        "displayDelta": "+10",
        "displayDeltaPercentage": "+100%",
        "displayValue": "10",
        "regression": true,
        "value": 10,
      },
      null,
    ],
  },
  {
    "biggerIsBetter": null,
    "changed": false,
    "key": "webpack.moduleCount",
    "label": "Modules",
    "runs": [
      {
        "delta": 0,
        "deltaPercentage": 0,
        "deltaType": "NO_CHANGE",
        "displayDelta": "0",
        "displayDeltaPercentage": "0%",
        "displayValue": "100",
        "regression": null,
        "value": 100,
      },
      {
        "displayValue": "100",
        "value": 100,
      },
    ],
  },
  {
    "biggerIsBetter": null,
    "changed": true,
    "key": "webpack.cacheInvalidation",
    "label": "Cache Invalidation",
    "runs": [
      null,
      {
        "displayValue": "10%",
        "value": 10,
      },
    ],
  },
]
`;

exports[`report / addRowMetricData should add data with metric type 1`] = `
[
  {
    "biggerIsBetter": false,
    "changed": true,
    "key": "metric1",
    "label": "metric1",
    "runs": [
      {
        "delta": -10,
        "deltaPercentage": -9.09090909,
        "deltaType": "POSITIVE",
        "displayDelta": "-10B",
        "displayDeltaPercentage": "-9.09%",
        "displayValue": "100B",
        "regression": false,
        "value": 100,
      },
      {
        "displayValue": "110B",
        "value": 110,
      },
    ],
  },
  {
    "biggerIsBetter": false,
    "changed": true,
    "key": "metric2",
    "label": "metric2",
    "runs": [
      {
        "delta": 10,
        "deltaPercentage": 100,
        "deltaType": "HIGH_NEGATIVE",
        "displayDelta": "+10B",
        "displayDeltaPercentage": "+100%",
        "displayValue": "10B",
        "regression": true,
        "value": 10,
      },
      null,
    ],
  },
  {
    "biggerIsBetter": false,
    "changed": true,
    "key": "metric3",
    "label": "metric3",
    "runs": [
      null,
      {
        "displayValue": "10B",
        "value": 10,
      },
    ],
  },
  {
    "biggerIsBetter": false,
    "changed": false,
    "key": "metric4",
    "label": "metric4",
    "runs": [
      {
        "delta": 0,
        "deltaPercentage": 0,
        "deltaType": "NO_CHANGE",
        "displayDelta": "0B",
        "displayDeltaPercentage": "0%",
        "displayValue": "100B",
        "regression": undefined,
        "value": 100,
      },
      {
        "displayValue": "100B",
        "value": 100,
      },
    ],
  },
]
`;
